### PR TITLE
Generate CODEOWNERS and comp/README.md for components

### PR DESCRIPTION
**Pull request by djmitche**
### What does this PR do?

Generate .github/CODEOWNERS and comp/README.md based on the components defined in comp/*

### Motivation

Automation makes things simpler and encourages regularity!

### Additional Notes

This sort of inherently accomplishes some linting, but we could probably do more.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: tasks/pylauncher.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-"""
+"""!:
```
Comment (mx-psi) [inline tasks/pylauncher.py:1]: Is this intentional?

File: tasks/components.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        if b.team:
+            yield f'/{b.path} @DataDog/{b.team}'
+        for c in b.components:
+            if c.team != b.team:
+                different_components.append(c)
+    for c in different_components:
+        if c.team:
+            yield f'/{c.path} @DataDog/{c.team}'
+
+    # drop lines from the existin codeowners until "# END COMPONENTS"
```
Comment (mx-psi) [inline tasks/components.py:138]:
```suggestion
    # drop lines from the existing codeowners until "# END COMPONENTS"
```

File: tasks/components.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            yield f'### [{c.path}](https://pkg.go.dev/{pkg_root}{c.path})'
+            yield ''
+            if c.team != b.team:
+                yield f'*Datadog Team*: {c.team}'
+                yield ''
+            yield c.doc
+
+
+def make_codeowners(codeowners_lines, bundles):
+    codeowners_lines = codeowners_lines.__iter__()
```
Comment (mx-psi) [inline tasks/components.py:114]: Why do you need to do this?

File: tasks/components.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            if current != components_md:
+                print(f"** {filename} differs")
+                ok = False
+                fixable = True
+
+    # Check .github/CODEOWNERS
+    filename = ".github/CODEOWNERS"
+    with open(filename, "r") as f:
+        current = f.read()
+    codeowners = '\n'.join(make_codeowners(current.split('\n'), bundles))
```
Comment (mx-psi) [inline tasks/components.py:173]:
```suggestion
    codeowners = '\n'.join(make_codeowners(current.splitlines(), bundles))
```

File: tasks/components.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        for c in b.components:
+            if c.team != b.team:
+                different_components.append(c)
+    for c in different_components:
+        if c.team:
+            yield f'/{c.path} @DataDog/{c.team}'
+
+    # drop lines from the existin codeowners until "# END COMPONENTS"
+    for line in codeowners_lines:
+        line = line.rstrip("\n")
```
Comment (mx-psi) [inline tasks/components.py:140]: (If you go with `splitlines`)
```suggestion
```

File: tasks/components.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        if line == "# END COMPONENTS":
+            yield line
+            break
+
+    # pass through the rest of the file
+    for line in codeowners_lines:
+        yield line.rstrip("\n")
```
Comment (mx-psi) [inline tasks/components.py:147]: (if you go with `splitlines`)
```suggestion
        yield line
```

File: tasks/components.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                yield ''
+            yield c.doc
+
+
+def make_codeowners(codeowners_lines, bundles):
+    codeowners_lines = codeowners_lines.__iter__()
+
+    # pass through the codeowners lines up to and including "# BEGIN COMPONENTS"
+    for line in codeowners_lines:
+        line = line.rstrip("\n")
```
Comment (mx-psi) [inline tasks/components.py:118]: (if you go with `splitlines`)
```suggestion
```

File: tasks/components.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            yield line
+            break
+
+    # pass through the rest of the file
+    for line in codeowners_lines:
+        yield line.rstrip("\n")
+
+
+@task
+def lint_components(ctx, fix=False):
```
Comment (mx-psi) [inline tasks/components.py:151]: Could you add a docstring for this task? Invoke can use it for CLI help messages

File: tasks/pylauncher.py (hunk 1 above)
Comment (djmitche) [reply tasks/pylauncher.py:1]: nope, looks like vim-spam :)

File: tasks/components.py (hunk 3 above)
Comment (djmitche) [reply tasks/components.py:114]: I call `next(..)` explicitly, so I need an iterator and not a list.

Comment (djmitche) [reply tasks/components.py:114]: Sorry, that was in an earlier version -- but, the `for` loops are meant to iterate sequentially, not to each begin at the beginning of the list.

**Review by mx-psi — APPROVED** [review]

**Review by hush-hush — CHANGES_REQUESTED** [review]

**Review by hush-hush — APPROVED** [review]

File: tasks/components.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        current = f.read()
+    codeowners = '\n'.join(make_codeowners(current.splitlines(), bundles))
+    if fix:
+        with open(".github/CODEOWNERS", "w") as f:
+            f.write(codeowners)
+    else:
+        if current != codeowners:
+            print(f"** {filename} differs")
+            ok = False
+            fixable = True
```
Comment (mx-psi) [inline tasks/components.py:188]:
```suggestion
    elif current != codeowners:
        print(f"** {filename} differs")
        ok = False
        fixable = True
```

File: tasks/components.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                yield ''
+            yield c.doc
+
+
+def make_codeowners(codeowners_lines, bundles):
+    codeowners_lines = codeowners_lines.__iter__()
+
+    # pass through the codeowners lines up to and including "# BEGIN COMPONENTS"
+    for line in codeowners_lines:
+        line = line
```
Comment (mx-psi) [inline tasks/components.py:119]:
```suggestion
```

File: tasks/components.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+        for c in b.components:
+            if c.team != b.team:
+                different_components.append(c)
+    for c in different_components:
+        if c.team:
+            yield f'/{c.path} @DataDog/{c.team}'
+
+    # drop lines from the existing codeowners until "# END COMPONENTS"
+    for line in codeowners_lines:
+        line = line
```
Comment (mx-psi) [inline tasks/components.py:141]:
```suggestion
```

**Comment by djmitche** [comment]
```
/usr/local/rvm/gems/ruby-2.7.5/gems/bundler-1.17.3/lib/bundler/resolver.rb:287:in `block in verify_gemfile_dependencies_are_found!': Could not find gem 'kitchen-azurerm (~> 1.10.5)' in any of the gem sources listed in your Gemfile. (Bundler::GemNotFound)
```
looks like the kitchen failures are unrelated to this.

**Pull request merged**